docs/man/man7/serc.7
.\" Process this file with
.\" groff -man -Tascii serc.7
.\"
.TH LIBSERC 7 2014.11.01 "" "SoftSprocket libserc"
.SH NAME
array_to_auto_string
array_to_set
char_ptr
create_array
create_id_tuple
create_literal_value
create_map
create_sequence
delete_document
delete_id_tuple
delete_literal_value
delete_storage_unit
document_to_auto_string
double_ptr
id_tuple_to_auto_string
int_ptr
literal_to_auto_string
map_to_auto_string
parseable_buffer_from_string
serc_document_to_auto_string
serc_parse_buffer
serc_read_file
set_to_auto_string
storage_unit_named_float
storage_unit_named_integer
storage_unit_named_char
storage_unit_named_string
storage_unit_named_heredoc
storage_unit_named_array
storage_unit_named_sequence
storage_unit_named_set
storage_unit_named_map
storage_unit_ptr
str_ptr
str_return_value
  \- library for processing Serc data serialization language 
.SH SYNOPSIS
.nf
.B #include <softsprocket/serc.h>
.B #include <softsprocket/container.h>
.sp
.B char* array_to_auto_string (auto_array* aa, auto_string* str);
.br
.B set* array_to_set (auto_array* aa);
.br
.B char* char_ptr (char value);
.br
.B auto_array* create_array (void* value);
.br
.B id_tuple* create_id_tuple (char* id, void* data, id_tuple_type type);
.br
.B literal_value* create_literal_value (literal_value_type type, void* data);
.br
.B hash_table* create_map (char* key, storage_unit* value);
.br
.B auto_array* create_sequence (hash_table* value);
.br
.B void delete_document (void* v);
.br
.B void delete_id_tuple (void* v);
.br
.B void delete_literal_value (void* v);
.br
.B void delete_storage_unit (void* v);
.br
.B char* document_to_auto_string (hash_table* document, auto_string* str);
.br
.B long double* double_ptr (long double value);
.br
.B char* id_tuple_to_auto_string (id_tuple* it, auto_string* str);
.br
.B long int* int_ptr (long int value);
.br
.B char* literal_to_auto_string (literal_value* lit, auto_string* str);
.br
.B char* map_to_auto_string (hash_table* ht, auto_string* str);
.br
.B char* parseable_buffer_from_string (char* str, size_t* buffer_size);
.br
.B auto_string* serc_document_to_auto_string (hash_table* document);
.br
.B hash_table* serc_parse_buffer (char* buffer, size_t buffer_size);
.br
.B char* serc_read_file (char* fname, size_t* buffer_size);
.br
.B char* set_to_auto_string (set* s, auto_string* str);
.br
.B storage_unit* storage_unit_named_float (long double value, char* name);
.br
.B storage_unit* storage_unit_named_integer (long int value, char* name);
.br
.B storage_unit* storage_unit_ptr (stored_type type, void* value);
.br
.B char* str_ptr (char* value);
.br
.B char* str_return_value (char* str);
.fi
.sp
Link with \fI\-lserc\fP \fI\-lsscont\fP.
.sp
.SH DESCRIPTION
Libserc is the reference library for the Serc data serialization language. Its functions
allow the programmer to parse or serialize C data types. For information on Serc see http://www.softsprocket.com/
.br
.sp
.nf

The enumerations and types used by libserc:

typedef enum {
        LV_STRING,
        LV_INTEGER,
        LV_FLOAT,
        LV_HEREDOC,
        LV_CHARACTER,
        LV_ARRAY
} literal_value_type;

typedef struct {
        literal_value_type type;
        void* value;
} literal_value;

typedef enum {
        TT_STRING,
        TT_HEREDOC,
        TT_INTEGER,
        TT_FLOAT,
        TT_CHARACTER,
        TT_ARRAY,
        TT_SET,
        TT_SEQUENCE,
        TT_MAP
} id_tuple_type;

typedef struct {
        id_tuple_type type;
        char* id;
        void* data;
} id_tuple;

typedef enum {
        ST_ID_TUPLE,
        ST_LITERAL_VALUE,
        ST_SUBDOC,
        ST_SET
} stored_type;

typedef struct {
        stored_type type;
        union {
                id_tuple* tuple;
                literal_value* literal;
                hash_table* map;
                set* s;
        };
} storage_unit;

.fi
.br
.sp
char* array_to_auto_string (auto_array* aa, auto_string* str);
.br
.in +4n
converts a Serc Array, stored as an auto_array, to an auto_string.
auto_string is from libsscont.
.br
aa - the auto_array to convert
.br
str - a pointer to the auto_string that will contain the converted array
.br
returns - a pointer to the nul terminated buf element of the auto_string or NULL in the event of an error
.br
.in
.sp
set* array_to_set (auto_array* aa);
.br
.in +4n
creates a set on the heap from the data of an auto_array of storage_units.
and deletes the auto_array. A Serc Set is represented by a set.
set is from libsscont
.br
aa - an auto_array of storage_unit pointers
.br
returns - a pointer to a set or NULL in the event of an error
.br
.in
.sp
char* char_ptr (char value);
.br
.in +4n
 allocates a char on the heap and assigns a value to it.
.br
 value - the value to assign 
.br
 returns -  a pointer to a char or NULL in the event of an error
.br
.in
.sp
auto_array* create_array (void* value);
.br
.in +4n
creates an auto_array on the heap and adds a value to it. A Serc Array or Set can be
represented by an auto_array. In both cases the object stored should be a storage_unit.
auto_array is from libsscont
.br
value - a pointer to a value to store or NULL in the event of an error
.br
a - pointer to an auto_array
.br
.in
.sp
id_tuple* create_id_tuple (char* id, void* data, id_tuple_type type);
.br
.in +4n
creates a id_tuple object on the heap.
.br
id - the name of the tuple
.br
data - a pointer to the value of the tuple 
.br
type - the type of the object
.br
returns - a pointer to the id_tuple or NULL in the event of an error
.br
.in
.sp
literal_value* create_literal_value (literal_value_type type, void* data);
.br
.in +4n
creates a literal_value object on the heap.
.br
type - the type of the object
.br
data - a pointer to the value of the literal
.br
returns - a pointer to the literal_value or NULL in the event of an error
.br
.in
.sp
hash_table* create_map (char* key, storage_unit* value);
.br
.in +4n
creates a hash_table on the heap and stores a value in it.
It is expected that the storage_unit contains an id_tuple
and its namewill be the key.
hash_table is from libsscont
.br
key - the key to store value under
.br
value - the value to be stored
.br
returns -  a pointer to a hash_table or NULL in the event of an error
.br
.in
.sp
auto_array* create_sequence (hash_table* value);
.br
.in +4n
creates an auto_array on the heap and adds a hash_table pointer to it.
A Serc sequence is reprsented by an auto_array. The hash_table represents
a Serc document or subdocument.
auto_array is from libsscont
.br
value - a hash_table to store
.br
returns - a pointer to an auto_array or NULL in the event of an error
.br
.in
.sp
void delete_document (void* v);
.br
.in +4n
free the memory from a Serc document (hash_table of storage_unit types).
Parameter is a void* to match libsscont hash_table_delete function pointer specification
but a hash_table* is expcted.
.br
v - the document to free
.br
.in
.sp
void delete_id_tuple (void* v);
.br
.in +4n
free the memory from a Serc id_tuple. 
Parameter is void* for use as function pointer to libsscont delete functions
but an id_tuple* is the expected type.
.br
v - the id_tuple to free
.br
.in
.sp
void delete_literal_value (void* v);
.br
.in +4n
free the memory from a Serc literal_value.
Parameter is void* for use as function pointer to libsscont delete functions
but a literal_value* is the expected type.
.br
v - the storage_unit to free
.br
.in
.sp
void delete_storage_unit (void* v);
.br
.in +4n
free the memory from a Serc storage_unit.
Parameter is void* for use as function pointer to libsscont delete functions
but a storage_unit* is the expected type.
.br
v - the storage_unit to free
.br
.in
.sp
char* document_to_auto_string (hash_table* document, auto_string* str);
.br
.in +4n
 converts a serc document stored as a hash_table into an auto_string.
 hash_table and auto_string are from libsscont
.br
 document - a hash_table containing the serc document elements
.br
 str - a pointer to an auto_str that will contain the converted document
.br
 returns - a pointer to the nul terminated buf element of the auto_string or NULL in the event of an error
.br
.in
.sp
long double* double_ptr (long double value);
.br
.in +4n
allocates a long double on the heap and assigns a value to it.
.br
value - the value to assign 
.br
returns - a pointer to a long double or NULL in the event of an error
.br
.in
.sp
char* id_tuple_to_auto_string (id_tuple* it, auto_string* str);
.br
.in +4n
converts a Serc id_tuple to an auto_string.
.br
it - the id_tuple to convert
.br
str - a pointer to the auto_string that will contain the converted id_tuple
.br
returns - a pointer to the nul terminated buf element of the auto_string or NULL in the event of an error
.br
.in
.sp
long int* int_ptr (long int value);
.br
.in +4n
allocates a long int on the heap and assigns a value to it.
.br
value - the value to assign 
.br
returns - a pointer to a long int or NULL in the event of an error
.br
.in
.sp
char* literal_to_auto_string (literal_value* lit, auto_string* str);
.br
.in +4n
converts a Serc literal_value to an auto_string.
.br
lit - the literal_value to convert
.br
str - a pointer to the auto_string that will contain the converted literal_value
.br
returns - a pointer to the nul terminated buf element of the auto_string or NULL in the event of an error
.br
.in
.sp
char* map_to_auto_string (hash_table* ht, auto_string* str);
.br
.in +4n
converts a hash_table representing a Serc map to an auto_string.
.br
ht - the hash_table to convert
.br
str - a pointer to the auto_string that will contain the converted map 
.br
returns - a pointer to the nul terminated buf element of the auto_string or NULL in the event of an error
.br
.in
.sp
char* parseable_buffer_from_string (char* str, size_t* buffer_size);
.br
.in +4n
convert a nul terminated c string into a double terminated parse buffer.
Makes a copy of the string
.br
str - a c string containing the Serc data
.br
buffer_size - an int pointer for the return buffer size
.br
returns - the parseable buffer or NULL in the event of an error
.br
.in
.sp
auto_string* serc_document_to_auto_string (hash_table* document);
.br
.in +4n
convenience function - creates an auto_string and calls document_to_auto_string.
auto_string and hash_table are from libsscont
.br
document - the Serc document stored as a hash_table
.br
returns - a pointer to an auto_string containing the document or NULL in the eventof an error
.br
.in
.sp
hash_table* serc_parse_buffer (char* buffer, size_t buffer_size);
.br
.in +4n
parses a Serc document into a form useable by a C program.
The buffer must terminated with two nul bytes.
hash_table is from libsscont
.br
buffer -  the string to parse
.br
buffer_size - the size of the buffer
.br
returns - a pointer to hash_table representing the Serc document or NULL in the event of an error
.br
.in
.sp
char* serc_read_file (char* fname, size_t* buffer_size);
.br
.in +4n
reads a file and converts it to a buffer suitable for Serc parse buffer
.br
fname - the path of the file to read
.br
point - to an int which will be populated with the returned buffer size
.br
returns - buffer containing the Serc document to be parsed or NULL in the eventof an error
.br
.in
.sp
char* set_to_auto_string (set* s, auto_string* str);
.br
.in +4n
converts a Serc set, stored as an libsscont set, to an auto_string.
.br
s - the set to convert
.br
str - a pointer to the auto_string that will contain the converted set
.br
returns - a pointer to the nul terminated buf element of the auto_string or NULL in the event of an error
.br
.in
.sp
storage_unit* storage_unit_named_float (long double value, char* name);
.br
.in +4n
creates a storage_unit of type ST_ID_TUPLE that contains an id_tuple* of
type TT_FLOAT. A convenience function that allocates the memory required
for the id, name and the tuple and makes the appropriate assignments.
A copy is made of the string name.
.br
value - the value to assign the id_tuple
.br
name - the name to assign the tuple
.br
returns - a storage_unit* or NULL if an error occurs
.br
.in
.sp
storage_unit* storage_unit_named_integer (long int value, char* name);
.br
.in +4n
creates a storage_unit of type ST_ID_TUPLE that contains an id_tuple* of
type TT_INTEGER. A convenience function that allocates the memory required
for the id, name and the tuple and makes the appropriate assignments.
A copy is made of the string name.
.br
value - the value to assign the id_tuple
.br
name - the name to assign the tuple
.br
returns - a storage_unit* or NULL if an error occurs
.br
.in
.sp
storage_unit* storage_unit_named_char (char value, char* name);
.br
.in +4n
creates a storage_unit of type ST_ID_TUPLE that contains an id_tuple* of
type TT_CHARACTER. A convenience function that allocates the memory required
for the id, name and the tuple and makes the appropriate assignments.
A copy is made of the string name.
.br
value - the value to assign the id_tuple
.br
name - the name to assign the tuple
.br
returns - a storage_unit* or NULL if an error occurs
.br
.in
.sp
storage_unit* storage_unit_named_string (char* value, char* name);
.br
.in +4n
creates a storage_unit of type ST_ID_TUPLE that contains an id_tuple* of
type TT_STRING. A convenience function that allocates the memory required
for the id, and the tuple and makes the appropriate assignments.
A copy is made of the string value. A copy is made of the string name.
.br
value - the value to assign the id_tuple
.br
name - the name to assign the tuple
.br
returns - a storage_unit* or NULL if an error occurs
.br
.in
.sp
storage_unit* storage_unit_named_heredoc (char* value, char* name);
.br
.in +4n
creates a storage_unit of type ST_ID_TUPLE that contains an id_tuple* of
type TT_HEREDOC. A convenience function that allocates the memory required
for the id, name and the tuple and makes the appropriate assignments.
The char* value should be on the heap (no copy is made). A copy is made of the string name.
.br
value - the value to assign the id_tuple
.br
name - the name to assign the tuple
.br
returns - a storage_unit* or NULL if an error occurs
.br
.in
.sp
storage_unit* storage_unit_named_array (auto_array* value, char* name);
.br
.in +4n
creates a storage_unit of type ST_ID_TUPLE that contains an id_tuple* of
type TT_ARRAY. A convenience function that allocates the memory required
for the id, name and the tuple and makes the appropriate assignments.
The auto_array* value should be on the heap (no copy is made). A copy is made of the string name.
.br
value - the value to assign the id_tuple
.br
name - the name to assign the tuple
.br
returns - a storage_unit* or NULL if an error occurs
.br
.in
.sp
storage_unit* storage_unit_named_set (set* value, char* name);
.br
.in +4n
creates a storage_unit of type ST_ID_TUPLE that contains an id_tuple* of
type TT_SET. A convenience function that allocates the memory required
for the id, name and the tuple and makes the appropriate assignments.
The set* value should be on the heap (no copy is made). A copy is made of the string name.
.br
value - the value to assign the id_tuple
.br
name - the name to assign the tuple
.br
returns - a storage_unit* or NULL if an error occurs
.br
.in
.sp
storage_unit* storage_unit_named_sequence (auto_array* value, char* name);
.br
.in +4n
creates a storage_unit of type ST_ID_TUPLE that contains an id_tuple* of
type TT_SEQUENCE. A convenience function that allocates the memory required
for the id, name and the tuple and makes the appropriate assignments.
The auto_array* value should be on the heap (no copy is made). A copy is made of the string name.
.br
value - the value to assign the id_tuple
.br
name - the name to assign the tuple
.br
returns - a storage_unit* or NULL if an error occurs
.br
.in
.sp
storage_unit* storage_unit_named_map (hash_table* value, char* name);
.br
.in +4n
creates a storage_unit of type ST_ID_TUPLE that contains an id_tuple* of
type TT_MAP. A convenience function that allocates the memory required
for the id, name and the tuple and makes the appropriate assignments.
The hash_table* value should be on the heap (no copy is made). A copy is made of the string name.
.br
value - the value to assign the id_tuple
.br
name - the name to assign the tuple
.br
returns - a storage_unit* or NULL if an error occurs
.br
.in
.sp
storage_unit* storage_unit_ptr (stored_type type, void* value);
.br
.in +4n
allocates a storage_unit on the heap and assigns a value and type to it.
.br
type - the type to assign
.br
value - the value to assign 
.br
returns - a pointer to a storage_unit or NULL in the event of an error
.br
.in
.sp
char* str_ptr (char* value);
.br
.in +4n
allocates a char* on the heap and coppies a value to it.
.br
value - the value to assign 
.br
returns - a char* or NULL in the event of an error
.br
.in
.sp
char* str_return_value (char* str);
.br
.in +4n
copies a string removing quotations from the front and back if they exist.
.br
str - the string to copy or NULL in the event of an error
.br
returns - a string
.br
.in
.in
.SH BUGS
No known bugs.
.SH AUTHOR
Greg Martin <greg@softsprocket.com>
.SH "SEE ALSO"
.BR hash_table (3),
.BR set (3),
.BR auto_string (3),
.BR auto_array (3)




src/serc.c
#include "serc.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <float.h>
#include <softsprocket/container.h>
#include <softsprocket/debug_utils.h>


char* str_return_value (char* str) {
	char* pstr;
	if (str[0] == '"') {
		pstr = &str[1];
	} else {
		pstr = &str[0];
	}
	
	int len = strlen (pstr);

	if (pstr[len - 1] == '"') {
		pstr[len - 1] = '\0';
		--len;
	}
	
	char* rv = malloc (len + 1);

	if (rv == NULL) {
		perror ("malloc");
		return NULL;
	}

	strcpy (rv, pstr);

	return rv;
}

literal_value* create_literal_value (literal_value_type type, void* data) {
	literal_value* dv = malloc (sizeof (literal_value));
	if (dv == NULL) {
		PERR ("malloc");
		return NULL;
	}

	dv->type = type;
	dv->value = data;
	return dv;
}

auto_array* create_array (void* value) {
	auto_array* aa = auto_array_create (1024);
	if (aa == NULL) {
		PMSG ("auto_array");
		return NULL;
	}

	auto_array_add (aa, value);

	return aa;
}

id_tuple* create_id_tuple (char* id, void* data, id_tuple_type type) {
	id_tuple* it = malloc (sizeof (id_tuple));
	if (it == NULL) {
		PERR ("malloc");
		return NULL;
	}

	it->id = id;
	it->data = data;
	it->type = type;

	return it;
}

int long_int_equals (void* this, void* other) {
	long int *l = this;
	long int *r = other;
	
	return *l == *r;
}

int long_double_equals (void* this, void* other) {
	long double *l = this;
	long double *r = other;
	
	long double al = fabsl (*l);
	long double ar = fabsl (*r);

	long double aor1 = (al > 1.0) ? al : 1.0;

	long double max = (aor1 > ar) ? aor1 : ar;
	long double diff = fabsl (*l - *r) / max;

	return diff < LDBL_EPSILON;
}

int set_equals (void* this, void* other) {
	set* l = this;
	set* r = other;

	if (l->count != r->count) {
		return 0;
	}

	for (int i = 0; i < l->count; ++i) {
		storage_unit* lsu = set_get_item (l, i);

		int match = 0;

		for (int j = 0; j < r->count; ++j) {
			storage_unit* rsu = set_get_item (r, j);
			
			if (lsu->type != rsu->type) {
				continue;
			}

			if (lsu->type == ST_SET) {
				if ((match = set_equals (lsu->s, rsu->s))) {
					break;
				}

			} else {
				if (lsu->literal->type == LV_INTEGER) {
					if ((match = long_int_equals (lsu->literal->value, rsu->literal->value))) {
						break;
					}			
				} else {
					if ((match = long_double_equals (lsu->literal->value, lsu->literal->value))) {
						break;
					}
				}
			}
		}

		if (!match) {
			return 0;
		}
			
	}

	return 1;
}

int storage_unit_equals (void* this, void* other) {
	storage_unit* lsu = this;
	storage_unit* rsu = other;

	switch (lsu->type) {
		case ST_SET: {
			set* l = lsu->s;
			set* r = rsu->s;
			return set_equals (l, r);
		}
		case ST_LITERAL_VALUE:

			if (lsu->literal->type == LV_INTEGER) {
				long int* l = lsu->literal->value;
				long int* r = rsu->literal->value;
				return long_int_equals (l, r);
			} else {
				long double* l = lsu->literal->value;
				long double* r = rsu->literal->value;
				return long_double_equals (l, r);
			}
		case ST_ID_TUPLE:
		case ST_SUBDOC:
			break; // not currently implemented
	}

	return 0;	
}

set* array_to_set (auto_array* aa) {

	set* s = set_create (aa->count, storage_unit_equals);
	if (s == NULL) {
		PMSG ("set");
		return NULL;
	}

	set_add_items (s, aa->data, aa->count);

	auto_array_delete (aa, NULL);

	return s;
}

auto_array* create_sequence (hash_table* value) {
	auto_array* seq = auto_array_create (100);
	if (seq == NULL) {
		PMSG ("auto_array");
		return NULL;
	}

	auto_array_add (seq, value);

	return seq;
}

hash_table* create_map (char* key, storage_unit* value) {
	hash_table* ht = hash_table_create (100);
	if (ht == NULL) {
		PMSG ("hash_table");
		return NULL;
	}
	
	hash_table_put (ht, key, value);

	if (value->type != ST_ID_TUPLE) {
		free (key);
	}

	return ht;
}

long int* int_ptr (long int value) {
	long int* p = malloc (sizeof (long int));
	if (p == NULL) {
		PERR ("malloc");
		return NULL;
	}

	*p = value;

	return p;
}

long double* double_ptr (long double value) {
	long double* p = malloc (sizeof (long double));
	if (p == NULL) {
		PERR ("malloc");
		return NULL;
	}

	*p = value;

	return p;
}

char* char_ptr (char value) {
	char* p = malloc (sizeof (char));
	if (p == NULL) {
		PERR ("malloc");
		return NULL;
	}

	*p = value;

	return p;
}

char* str_ptr (char* value) {
	char* p = malloc (strlen (value) + 1);
	if (p == NULL) {
		PERR ("malloc");
		return NULL;
	}

	strcpy (p, value);

	return p;
}

storage_unit* storage_unit_ptr (stored_type type, void* value) {
	storage_unit* su = malloc (sizeof (storage_unit));
	if (su == NULL) {
		PERR ("malloc");
		return NULL;
	}

	su->type = type;

	if (type == ST_ID_TUPLE) {
		su->tuple = value;
	} else if (type == ST_SUBDOC) {
		su->map = value;
	} else if (type == ST_SET) {
		su->s = value;
	} else {
		su->literal = value;
	}

	return su;
}

char* array_to_auto_string (auto_array* aa, auto_string* str) {
	char* begin = "[";
	auto_string_append (str, begin);
	char* comma = ",";
	int use_comma = 0;

	for (int i = 0; i < aa->count; ++i) {

		if (use_comma) {
			auto_string_append (str, comma);
		}

		literal_value* lv = auto_array_get (aa, i);
		literal_to_auto_string (lv, str);

		use_comma = 1;

	}

	char* term = "]";
	auto_string_append (str, term);

	return str->buf;
}

char* set_to_auto_string (set* s, auto_string* str) {
	auto_string_append (str, "(");

	char* comma = ",";
	int use_comma = 0;

	for (size_t i = 0; i < s->count; ++i) {
		if (use_comma) {
			auto_string_append (str, comma);
		}

		storage_unit* su = set_get_item (s, i);
		if (su->type == ST_ID_TUPLE) {
			id_tuple_to_auto_string (su->tuple, str);
		} else if (su->type == ST_SET) {
			set_to_auto_string (su->s, str);
		} else {
			literal_to_auto_string (su->literal, str);
		}

		use_comma = 1;
	}

	auto_string_append (str, ")");
	return str->buf;
}

char* map_to_auto_string (hash_table* ht, auto_string* str) {
	auto_string_append (str, "{");

	char* comma = ",";
	int use_comma = 0;

	auto_array* keys = hash_table_keys (ht);

	for (int i = 0; i < keys->count; ++i) {
		if (use_comma) {
			auto_string_append (str, comma);
		}

		char* key = auto_array_get (keys, i); 
		storage_unit* su = hash_table_get (ht, key);
		auto_string_append (str, key);
		auto_string_append (str, ":");

		if (su->type == ST_ID_TUPLE) {
			id_tuple_to_auto_string (su->tuple, str);
		} else if (su->type == ST_LITERAL_VALUE) {
			literal_to_auto_string (su->literal, str);
		} else {
			document_to_auto_string (su->map, str);
		}

		use_comma = 1;
	}	

	auto_string_append (str, "}");
	
	auto_array_delete (keys, NULL);

	return str->buf;
}

char* literal_to_auto_string (literal_value* lit, auto_string* str) {
	int sz = 256;
	char* val = malloc (sz);
	if (val == NULL) {
		PERR ("malloc");
		return NULL;
	}

	int append_val = 1;

	int rv = 0;
	switch (lit->type) {
		case LV_STRING:
			while ((rv = snprintf (val, sz, "\"%s\"", (char*) lit->value)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}

			break;
		case LV_INTEGER:
			while ((rv = snprintf (val, sz, "%ld", *(long int*) lit->value)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}

			break;
		case LV_FLOAT:
			while ((rv = snprintf (val, sz, "%Lf", *(long double*) lit->value)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}

			break;
		case LV_HEREDOC:
			while ((rv = snprintf (val, sz, "```%s```", (char*) lit->value)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}

			break;
		case LV_CHARACTER:
			while ((rv = snprintf (val, sz, "'%c'", *(char*) lit->value)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}

			break;
		case LV_ARRAY:
			array_to_auto_string (lit->value, str);
			append_val = 0;
			break;
		default:
			free (val);
			PMSG ("Unknown literal value type");
			return NULL;
	}

	if (append_val) {
		auto_string_append (str, val);
	}

	free (val);

	return str->buf;	
}

char* id_tuple_to_auto_string (id_tuple* it, auto_string* str) {

	int sz = 256;
	char* val = malloc (sz);
	if (val == NULL) {
		PERR ("malloc");
		return NULL;
	}

	int rv = 0;
	switch (it->type) {
		case TT_STRING:
			while ((rv = snprintf (val, sz, "String %s=\"%s\"", it->id, (char*) it->data)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}
			auto_string_append (str, val);

			break;
		case TT_HEREDOC:
			while ((rv = snprintf (val, sz, "String %s=```%s```", it->id, (char*) it->data)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}
			auto_string_append (str, val);
			break;
		case TT_INTEGER:
			while ((rv = snprintf (val, sz, "Integer %s=%ld ", it->id, *(long int*) it->data)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}
			auto_string_append (str, val);
			break;
		case TT_FLOAT:
			while ((rv = snprintf (val, sz, "Float %s=%Lg ", it->id, *(long double*) it->data)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}
			auto_string_append (str, val);
			break;
		case TT_CHARACTER:
			while ((rv = snprintf (val, sz, "Character %s='%c' ", it->id, *(char*) it->data)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}
			auto_string_append (str, val);
			break;
		case TT_ARRAY:
			while ((rv = snprintf (val, sz, "Array %s=", (char*) it->id)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}
			auto_string_append (str, val);
			array_to_auto_string (it->data, str);
			//auto_string_append (str, " ");
			break;
		case TT_SET:
			while ((rv = snprintf (val, sz, "Set %s=", (char*) it->id)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}

			auto_string_append (str, val);
			set_to_auto_string (it->data, str);
			//auto_string_append (str, " ");
			break;
		case TT_SEQUENCE:
			while ((rv = snprintf (val, sz, "Sequence %s={", (char*) it->id)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}
			auto_string_append (str, val);
			auto_array* sq = it->data;
			char* comma = ",";
			int use_comma = 0;

			for (int i = 0; i < sq->count; ++i) {
				if (use_comma) {
					auto_string_append (str, comma);
				}
				hash_table* doc = auto_array_get (sq, i);
				document_to_auto_string (doc, str);
				use_comma = 1;
			}
			auto_string_append (str, "}");
			break;
		case TT_MAP:
			while ((rv = snprintf (val, sz, "Map %s=", (char*) it->id)) > sz) {
				sz = rv;
				char* tmp = realloc (val, sz);
				if (tmp == NULL) {
					PERR ("malloc");
					free (val);
					return NULL;
				}

				val = tmp;
			}
			auto_string_append (str, val);
			map_to_auto_string (it->data, str);
			break;
		default:
			PMSG ("Unknown tuple type");
			free (val);
			return NULL;
	}

	free (val);

	return str->buf;	
}

char* document_to_auto_string (hash_table* document, auto_string* str) {
	auto_array* keys = hash_table_keys (document);
	if (keys == NULL) {
		PERR ("auto_array");
		return NULL;
	}

	auto_string_append (str, "{");

	for (int i = 0; i < keys->count; ++i) {
		char* key = auto_array_get (keys, i);
		id_tuple* it = ((storage_unit*) hash_table_get (document, key))->tuple;
		id_tuple_to_auto_string (it, str);
	}

	auto_string_append (str, "}");

	auto_array_delete (keys, NULL);

	return str->buf;	
}

void delete_id_tuple (void* v) {
	id_tuple* tuple = v;
	switch (tuple->type) {
		case TT_STRING:
		case TT_HEREDOC:
		case TT_INTEGER:
		case TT_FLOAT:
		case TT_CHARACTER:
			free (tuple->data);
			break;
		case TT_ARRAY: {
				auto_array* aa = tuple->data;
				auto_array_delete (aa, delete_literal_value);
			}
			break;
		case TT_SET: {
				set* s = tuple->data;
				set_delete (s, delete_storage_unit);
			}
			break;
		case TT_SEQUENCE: {
				auto_array* seq = tuple->data;
				auto_array_delete (seq, delete_document);
			}
			break;
		case TT_MAP: {
				hash_table* ht = tuple->data;
				hash_table_delete (ht, delete_storage_unit);
			}
			break;
		default:
			PMSG ("Unknown id_tuple type");
			break;

	}

	free (tuple->id);
	free (tuple);
}

void delete_literal_value (void* v) {
	literal_value* literal = v;
	if (literal->type == LV_ARRAY) {
		auto_array_delete (literal->value, delete_literal_value);
	} else {
		free (literal->value);
	}

	free (literal);
}

void delete_storage_unit (void* v) {
	storage_unit* su = v;
	switch (su->type) {
		case ST_ID_TUPLE:
			delete_id_tuple (su->tuple);
			break;
		case ST_LITERAL_VALUE:
			delete_literal_value (su->literal);
			break;
		case ST_SUBDOC:
			delete_document (su->map);
			break;
		case ST_SET:
			set_delete (su->s, delete_storage_unit);
			break;
		default:
			PMSG ("Unknown storage unit type");
			break;
	}

	free (su);
}

void delete_document (void* v) {
	hash_table_delete (v, delete_storage_unit);
}


auto_string* serc_document_to_auto_string (hash_table* document) {
	auto_string* the_str = auto_string_create (32); 
	document_to_auto_string (document, the_str);

	return the_str;
}

